gnomad_qc/v4/assessment/calculate_per_sample_stats.py

```diff
@@ -389,16 +442,292 @@ def create_per_sample_counts_ht(
+          filter groups times the possible permutations of variant level stats. If
+          This number is too small, it might be better to use
+          `create_per_sample_counts_ht`.
+        - Sample IDs are mapped to the sample indices and the Table is exploded so that
+          each row is counts for a sample and a filter group. The number of rows in the
+          Table is approximately the number of samples times the number of rows in the
+          previous Table.
+        - Group the Table by sample to get a struct of the count of variants by sample
+          QC metric for each sample in each filter group.
+        - Add sample QC stats that are combinations of other stats.
+
+    :param ht: Input Table containing variant data. Must have multi-allelic sites split.
+    :return: Table containing per-sample variant counts.
+    """
+    # Get the number of filter groups, samples, and variants in the Table to determine
+    # a reasonable number of partitions to use for the aggregated filter group HT, and
+    # the number of random groups needed to have approximately the desired partitions.
+    n_filter_groups = len(hl.eval(ht.filter_group_meta))
+    n_samples = hl.eval(hl.len(ht.samples))
+    n_variants = hl.eval(ht.count())
+    # 2^3 = 8 because there are about 3 variant level stats: singletons (singleton_ti
+    # and singleton_tv will never overlap, and one of them will always also be a
+    # singleton), transition/transversion, insertion/deletion.
+    n_filter_permutations = n_filter_groups * 8
+    # Split the Table into approximately 10,000 variants per partition.
+    n_mem_partitions = max(int(n_variants / 10000), n_filter_permutations)
+    # Compute the number of random groups needed to have approximately n_mem_partitions.
+    n_rand_groups = max(int(n_mem_partitions / n_filter_permutations), 1)
+    logger.info(
+        "Aggregating counts for each sample in each filter group...\n"
+        "\tNumber of samples: %d\n"
+        "\tNumber of variants: %d\n"
+        "\tNumber of filter groups: %d\n"
+        "\tApproximate number of filter permutations: %d\n\n"
+        "\tCalculated number of partitions to improve memory usage: %d\n"
+        "\tNumber of random groups to have approximately the above partitions: %d",
+        n_samples,
+        n_variants,
+        n_filter_groups,
+        n_filter_permutations,
+        n_mem_partitions,
+        n_rand_groups,
+    )
+
+    # Group by filter groups and the random group to get the count of variants matching
+    # the filter group and random group for each sample.
+    ht = ht.annotate(_rand_group=hl.rand_int32(n_rand_groups))
+    ht = ht.group_by("filter_groups", "_rand_group").aggregate(
+        counts=hl.struct(
+            **{
+                k: hl.agg.explode(lambda x: hl.agg.counter(x), v)
+                for k, v in ht.sample_idx_by_stat.items()
+            }
+        )
+    )
```

Comment:
It's really just to increase the number of partitions since `filter_groups` might be smaller than `n_mem_partitions` and then it hits memory errors because there are too few partitions (since when grouping by `filter_groups` the total number of rows will be the number of `filter_groups`, and therefore a max number of partitions that it can be is `filter_groups`)

Reply:
added to the docstring, let me know if it makes sense